### PR TITLE
✨ Implement "upgrade plan" support

### LINKED ISSUE #129: Support clusterctl upgrade plan
**User Story**

As a user/operator I would like to determine the current CAPI* versions being used on the management cluster as well as if there is an upgrade available using `clusterctl upgrade plan` so that I don't have to take as long to manually figure this out probing deployments or other manual means.

**Detailed Description**
Only support `clusterctl upgrade plan` to determine current and upgradeable versions for the mangement cluster.  `clusterctl upgrade apply` out of scope for this feature request.  Currently if you run this against a cluster which has CAPI and the provider installed, you get this:

```
Checking new release availability...
Error: invalid management cluster: there should a core provider, found 0
```

/kind feature

### PR DESCRIPTION
**Pull request by Fedosin**
**What this PR does / why we need it**:

This PR implements `upgrade plan` support for the plugin

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #129

### CODE REVIEW

File: cmd/plugin/cmd/init_test.go [hunk 1]
```diff
@@ -375,100 +372,6 @@ func generateCAPIOperatorDeployment(name, namespace string) *appsv1.Deployment {
 	}
 }
```
Comment (Danil-Grigorev) [inline cmd/plugin/cmd/init_test.go:373]: One question - this test removed, is it covered by upgrade plan tests? Shouldn't we leave both for clarity?

File: cmd/plugin/cmd/upgrade_plan_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+						if err != nil {
+							return false, err
+						}
+
+						if provider.GetSpec().FetchConfig == nil || provider.GetSpec().FetchConfig.URL != tt.customURL {
+							return false, nil
+						}
+
+						return true, nil
+					}, waitShort).Should(BeTrue())
```
Comment (Danil-Grigorev) [inline cmd/plugin/cmd/upgrade_plan_test.go:159]: This format makes things a bit more tidy IMO.
```suggestion
					g.Eventually(func(g g.Gomega) error {
						provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
						g.Expect(err).ToNot(HaveOccurred())
						g.Expect(provider.GetSpec().FetchConfig != nil && provider.GetSpec().FetchConfig.URL == tt.customURL).To(BeTrue())
					}, waitShort).Should(Succeed())
```

**Review by Danil-Grigorev — COMMENTED** [review]
Looks good to me, just some questions/suggestions/nits

**Review by alexander-demicev — APPROVED** [review]
/approve

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm
/hold 

To give time to resolve comments from @Danil-Grigorev

File: cmd/plugin/cmd/upgrade_plan.go [hunk 3]
```diff
@@ -89,80 +119,329 @@ func init() {
...
+	// Get IPAM Providers.
+	var ipamProviderList operatorv1.IPAMProviderList
+
+	if err := client.List(ctx, &ipamProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of ipam providers from the server: %w", err)
+	}
+
+	for i := range ipamProviderList.Items {
+		genericProviders = append(genericProviders, &ipamProviderList.Items[i])
+	}
```
Comment (Danil-Grigorev) [inline cmd/plugin/cmd/upgrade_plan.go:416]: Please add runtime extension providers here too

**Review by Danil-Grigorev — COMMENTED** [review]
One question to runtime extension provider addition, otherwise lgtm.

File: cmd/plugin/cmd/upgrade_plan.go (hunk 3 above)
Comment (Fedosin) [reply cmd/plugin/cmd/upgrade_plan.go:416]: working on this!

**Comment by Danil-Grigorev** [comment]
/lgtm

Will leave hold in case Mike wants to re-add removed test, otherwise everything looks good. Thanks!

**Pull request merged**